rust/connlib/tunnel/src/peer.rs

```diff
@@ -20,25 +20,46 @@ type ExpiryingResource = (ResourceDescription, Option<DateTime<Utc>>);
 // is 30 seconds. See resolvconf(5) timeout.
 const IDS_EXPIRE: std::time::Duration = std::time::Duration::from_secs(60);
 
-pub struct Peer<TId, TTransform> {
-    allowed_ips: IpNetworkTable<()>,
+pub struct Peer<TId, TTransform, TResource> {
+    // TODO: we should refactor this
+    // in the gateway-side this means that we are explicit about ()
+    // maybe duping the Peer struct is the way to go
+    pub allowed_ips: IpNetworkTable<TResource>,
     pub conn_id: TId,
     pub transform: TTransform,
 }
 
-impl<TId, TTransform> Peer<TId, TTransform>
+impl<TId, TTransform> Peer<TId, TTransform, HashSet<ResourceId>>
 where
     TId: Copy,
     TTransform: PacketTransform,
+{
+    pub(crate) fn insert_id(&mut self, ip: &IpNetwork, id: &ResourceId) {
+        if let Some(resources) = self.allowed_ips.exact_match_mut(*ip) {
```

Comment:
You could use `get_or_insert` here to populate the hashset if it is not yet there: https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert

Reply:
ah! nice

Reply:
the problem with this is that when using `get_or_insert` modifies the returned `Option`(which is an `Option` not a `&mut Option`) so the underlying value was not modified into a `HashSet`